### PR TITLE
fix(metrics): explicit type to single_metric ctx manager

### PR DESCRIPTION
**Pull request by whardier**
**Issue #, if available:**

* [Maintenance: address top papercuts #857](https://github.com/awslabs/aws-lambda-powertools-python/issues/857)

## Description of changes:

Simply update type return for `metrics.metric.single_metric`.

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [ ] Update tests
* [ ] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws_lambda_powertools/metrics/metric.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def add_metric(self, name: str, unit: Union[MetricUnit, str], value: float) -> N
 
 
 @contextmanager
-def single_metric(name: str, unit: MetricUnit, value: float, namespace: Optional[str] = None):
+def single_metric(name: str, unit: MetricUnit, value: float, namespace: Optional[str] = None) -> Iterator[SingleMetric]:
```
Comment (heitorlessa) [inline aws_lambda_powertools/metrics/metric.py:64]: TIL: An Iterator *could* be problematic if someone tries to iterate it: https://github.com/python/typeshed/issues/2772

What we could do instead is using `Generator[yield_value, NoSendType, NoReturnType]` resulting in `Generator[SingleMetric, None, None]`.

```suggestion
def single_metric(name: str, unit: MetricUnit, value: float, namespace: Optional[str] = None) -> Generator[SingleMetric, None, None]:
```

File: aws_lambda_powertools/metrics/metric.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import json
 import logging
 from contextlib import contextmanager
-from typing import Dict, Optional, Union
+from typing import Dict, Iterator, Optional, Union
```
Comment (heitorlessa) [inline aws_lambda_powertools/metrics/metric.py:4]: See reason at the bottom

```suggestion
from typing import Dict, Optional, Union, Generator
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
One minor change to use `Generator[SingleMetric, None, None]` due to a new issue discovered in Python typeshed when using `Iterator[]` in this context.

File: aws_lambda_powertools/metrics/metric.py (hunk 1 above)
Comment (heitorlessa) [reply aws_lambda_powertools/metrics/metric.py:64]: On a practical level, mypy wise, it leads to the same type result:

image

image

**Comment by heitorlessa** [comment]
I created a new project using VSCode and PyLance and I couldn't reproduce this issue, same with PyCharm so perhaps their InteliSense is smart by inspecting the yield type `SingleMetric`.

Nevertheless, [TIL why](https://github.com/python/typeshed/issues/2772) we can't use `Iterator[SingleMetric]` and instead should use `Generator[SingleMetric, None, None`.

For the sake of correctness regardless of what IDE one might use, I'm gonna commit the Generator suggestion to avoid issues and merge it.

**Review by heitorlessa — APPROVED** [review]

**Comment by whardier** [comment]
Interesting.  I poured over a ton of python code and standard libs and Iterator seemed like the best fit.  I will read up on this.

**Comment by heitorlessa** [comment]
Yep same. It seems to be a “new” thing

**Pull request merged**